fe/fe-core/src/main/java/org/apache/doris/nereids/glue/translator/PhysicalPlanTranslator.java

```diff
@@ -257,6 +283,210 @@ public PlanFragment visitPhysicalAggregate(
+        return inputPlanFragment;
+    }
+
+    /**
+     * Generate outputSlot based on groupBy and output information.
+     * The generated outputSlot needs to guarantee the order,
+     * first slotReference and then virtualSlotReference.
+     *
+     * First collect the slot and then collect the virtualSlot.
+     * 1. Put the columns in groupBy into the slot
+     * 2. Put the columns in the output into the slot
+     * 3. Put slot and virtualSlot into finalSLots in turn.
+     *
+     * eg: select sum(k2) grouping(k1) from t1 group by grouping sets((k1));
+     * 1. slots:[k1]
+     * 2. slots:[k1, k2]
+     * 3. finalSlots: [k1, k2, GROUPING_PREFIX_k1]
+     */
+    private List<Slot> genOutputSlots(
+            List<SlotReference> groupBySlot,
+            List<NamedExpression> outputExpressionList,
+            List<Slot> virtualSlots) {
+
+        Map<Boolean, List<Slot>> allSlots = groupBySlot.stream()
+                .collect(Collectors.groupingBy(VirtualSlotReference.class::isInstance,
+                        LinkedHashMap::new, Collectors.toList()));
+        List<Slot> nonVirtualSlots = allSlots.get(false);
+        Map<Boolean, List<NamedExpression>> allOutput = outputExpressionList.stream()
+                .collect(Collectors.groupingBy(VirtualSlotReference.class::isInstance,
+                        LinkedHashMap::new, Collectors.toList()));
+        List<NamedExpression> nonVirtualOutput = allOutput.get(false);
+
+        nonVirtualSlots.addAll(nonVirtualOutput.stream()
+                .filter(e -> !nonVirtualSlots.contains(e))
+                .filter(SlotReference.class::isInstance)
+                .map(NamedExpression::toSlot)
+                .collect(Collectors.toSet()));
+        List<Slot> finalSlots = Lists.newArrayList();
+        finalSlots.addAll(nonVirtualSlots);
+        finalSlots.addAll(virtualSlots);
+        return finalSlots;
+    }
+
+    /**
+     * Get all the used columns that appear in repeatNode according to the column of groupBy
+     * and the column of aggFunc in output.
+     *
+     * eg: select sum(k2) grouping(k1) from t1 group by grouping sets((k1));
+     * preExpressions: [k1, k2].
+     */
+    private List<Expression> genPreExpressions(
+            List<Expression> nonVirtualGroupByExpressions, List<NamedExpression> outputExpressionList) {
+        List<Expression> preExpressions = Lists.newArrayList();
+        preExpressions.addAll(nonVirtualGroupByExpressions);
+        Map<Boolean, List<NamedExpression>> allOutput = outputExpressionList.stream()
+                .collect(Collectors.groupingBy(VirtualSlotReference.class::isInstance,
+                        LinkedHashMap::new, Collectors.toList()));
+        List<NamedExpression> nonVirtualOutput = allOutput.get(false);
+        nonVirtualOutput.stream()
+                .filter(e -> !preExpressions.contains(e))
+                .forEach(preExpressions::add);
+        return preExpressions;
+    }
```

Comment:
ditto
```java
private List<Expression> genPreExpressions(
            List<Expression> nonVirtualGroupByExpressions, List<NamedExpression> outputExpressionList) {
    Set <Slot> allSlots = ImmutableSet.builder()
        .addAll(groupByExpressions.stream()
            .flatMap(groupExpr -> groupExpr.getInputSlots().stream()))
            .collect(ImmutableSet.toSet())
        .addAll(outputExpressions.stream()
            .flatMap(outputExpr -> outputExpr.getInputSlots().stream()))
            .collect(ImmutableSet.toSet())
        ).build();

    return allSlots.stream()
        .filter(slot -> !(slot instanceof VirtualSlotReference))
        .collect(ImmutableList.toImmutableList());
}
```

Reply:
Look seems the used slots in aggregate/repeat will use lots of time, so you can extract all used slot, and extract slots according to your needs.

```java
Set<Slot> usedSlots = extractUsedSlots(groupByExpressions, outputExpressions);

private List<Slot> getUsedSortedSlots(Set<Slot> allSlot);
private List<Expression> getUsedNonVirtualSlots(Set<Slot> allSlot);
```